File: #442/Solution.java
/**
 * https://leetcode.com/problems/find-all-duplicates-in-an-array/
 */

class Solution {
	public List<Integer> findDuplicates(int[] nums) {
		List<Integer> duplicates = new ArrayList<Integer>();
		Set<Integer> noDupes = new HashSet<Integer>();
        for (int i = 0; i < nums.length; i++){
            if(noDupes.add(nums[i]) == false) {
            	duplicates.add(nums[i]);
            }
        }
        return duplicates;
	}
}

File: #442/cpp/solution.cpp
/**
* https://leetcode.com/problems/find-all-duplicates-in-an-array/
* Runtime beats 69.91% of C++ submissions
* Memory Usage beats 90.00% of C++ submissions
*/
class Solution {
public:
    vector<int> findDuplicates(vector<int>& nums) {
        int size = nums.size();
        vector<int> dups;

        for(int i = 0; i < size;)
            if(nums[nums[i]-1] != nums[i])
                swap(nums[i], nums[nums[i]-1]);
            else i++;

        for(int i = 0;i < size;i++)
            if(nums[i]!=i+1) 
                dups.push_back(nums[i]);
                
        return dups;
    }
};